### ISSUE TITLE
Cannot Save Weights with Model Checkpoint

### ISSUE THREAD

**Issue by pGit1** [titu1994/Keras-NASNet#16]
This model as it stands is not useful for custom datasets since there is no easy way to save weights with Model Checkpoint. The layer names are too long. Im hacking my way through trying to figure out a solution but the fact that weights are saveable (at least in Keras 2.1.3) should come with a disclaimer. I realize it is an HD5 issue but still extremely frustrating. 

ModelCheckpoint is already set to save weights only and save best only.

**Comment by titu1994** [comment]
I don't understand how that actually works. I did not make any changes to the weight naming scheme, yet I was able to save weights without an issue, even for the largest NASNet model. 

Then why is it that only when fine-tuning such an issue arises? Are you adding a large number of layers to the end of the network?

Technically speaking, if I could save weights with model.save_weights() then I should be able to set weights with Model checkpoint.

**Comment by pGit1** [comment]
>Technically speaking, if I could save weights with model.save_weights() then I should be able to set weights with Model checkpoint.

Yes you should. Looking at the Modelchekpoint Callback code it is literally doing a model.save_weights().

The error that keeps popping up is:
```
RuntimeError: Unable to create attribute (Object header message is too large)
```
This is supposed to be due to some 64KB header file limit, but Ive checked, the NASNet Mobile layers only take up about 32KB and some change when running this command:

```
np.asarray([layer.name.encode('utf8') for layer in model.layers]).nbytes
```
So this is extremely annoying.

All my hacks to this point have not worked. Ive tried modifying the `save_weights_to_hdf5_group`
directly to no avail.

> Then why is it that only when fine-tuning such an issue arises? Are you adding a large number of layers to the end of the network?

No. Just 10.

**Comment by pGit1** [comment]
Im closing this because I FINALLY found a hack that worked...but it is not all general and I found by just a step above random search and playing with my code.

**Comment by titu1994** [comment]
What's the fix? Even if it's non general it's useful in the case of NASNets. 

I remember someone else calculating this before as well, and there wasn't any discernable reason as to why the 64kb limit was being surpassed.

**Comment by pGit1** [comment]
Hi @titu1994 

Can I just post you some code? I had to modify the native Keras Modelcheckpoint callback and used some code from keras_contrib. "save_all_weights"

See my code below: Once I have this code I import the ModelCheckpoint from this file as opposed to the regular keras.callbacks version.

Look at lines, 319, 331. 417-424

[NASNet_ModelCheckPoint.txt](https://github.com/titu1994/Keras-NASNet/files/2237367/NASNet_ModelCheckPoint.txt)

Importing ModelCheckpoint from here instead of regular model checkpoint allows one to use model checkpoint callback as usual. The key with the headerfile thing ended up being the weight names ended up being to long when using the `save_weights_to_hdf5_group` so I modified that part of the funtion directly by truncating weight name, but still letting them be unique with integers (as strings since they are part of a string name).

Hope this helps.
